[mlir][vector] Clean up VectorTransferOpInterface

matthias-springer opened the pull request:
- Better documentation.
- Rename interface methods: `source` -> `getSource`, `indices` -> `getIndices`, etc. to conform with MLIR naming conventions. A default implementation is not needed.
- Turn many interface methods into helper functions. Most of the previous interface methods were not meant to be overridden, and if some were overridden without others, the op would be have been broken.

dcaballe commented on mlir/include/mlir/Interfaces/VectorInterfaces.td:
```diff
@@ -46,246 +48,249 @@ def VectorUnrollOpInterface : OpInterface<"VectorUnrollOpInterface"> {
+        // is because there are only 2 transfer dimensions.
+        %0 = vector.transfer_read %arg1[%c3, %c3], %vf0
+            {permutation_map = affine_map<(d0, d1) -> (d0, d1)>}
+            : memref<?x?xvector<4x3xf32>>, vector<1x1x4x3xf32>
```
If this is coming from lit tests, we had some weird cases for vector element types that were supported but we thought they should be deprecated.

matthias-springer replied:
This is copied from `Vector/ops.mlir`. It would be nice to drop vector element types, that would simplify the ops a lot and we would also no longer have to distinguish between "vector type rank" and "transfer rank". I'm preparing a separate change for this.

dcaballe replied:
I think vector element types are needed in general because they enforce the alignment of the elements to the vector size boundary. This is similar to allocating a vector type in LLVM. Part of the complexity comes from how we apply the permutation maps to these memrefs so perhaps we could just be more restrictive there. We can limit permutation to simple cases that are well defined and see how things go...

dcaballe reviewed (approved):
Thank! LGTM @ pending comments

matthias-springer committed:
[mlir][vector] Clean up VectorTransferOpInterface

- Better documentation.
- Rename interface methods: `source` -> `getSource`, `indices` -> `getIndices`, etc. to conform with MLIR naming conventions. A default implementation is not needed.
- Turn many interface methods into helper functions. Most of the previous interface methods were not meant to be overridden, and if some of the would have been overridden without other, the op would be have been broken.
